### PR TITLE
Add the Adaptive Random Forest

### PR DESCRIPTION
**Pull request by jacobmontiel**
This PR covers the Adaptive Random Forest Classifier and Regressor

- Refactor ARF
- Protects the `pdf` method against `TypeError` and `OverflowError`
- Refactor the class structures to minimize the amount of duplicated code.

### CODE REVIEW

**Comment by jacobmontiel** [comment]
Note: Tests in Travis are failing due to a corner case for the classification matrix. The fix to this issue is in the SRP PR.

File: river/ensemble/test_adaptive_random_forest.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            n_models=3,   # Smaller ensemble than the default to avoid bottlenecks
+            seed=42
+        )
+    ]
+    for check in utils.estimator_checks.yield_checks(estimator)
+    # Skipping this test since shuffling features does impact the Adaptive Random Forest
+    if check.__name__ not in {'check_shuffle_features_no_impact'}
+])
+def test_check_estimator(estimator, check):
+    check(copy.deepcopy(estimator))
```
Comment (smastelini) [inline river/ensemble/test_adaptive_random_forest.py:36]: Please, add an empty line.

File: river/proba/gaussian.py [hunk 2]
```diff
@@ -55,7 +55,12 @@ def update(self, x, w=1.):
     def pdf(self, x):
         var = self._var.get()
         if var:
-            return math.exp((x - self.mu) ** 2 / (-2 * var)) / math.sqrt(math.tau * var)
+            try:
```
Comment (smastelini) [inline river/proba/gaussian.py:58]: Nice catch!

File: river/test_.py [hunk 3]
```diff
@@ -36,6 +36,8 @@ def get_all_estimators():
         compose.FuncTransformer,
         compose.Pipeline,
         compose.Grouper,
+        ensemble.AdaptiveRandomForestClassifier,
+        ensemble.AdaptiveRandomForestRegressor,
```
Comment (smastelini) [inline river/test_.py:40]: Should we shorten their names and follow the path paved by SRPClassifier?

File: river/tree/arf_hoeffding_tree_classifier.py [hunk 4]
```diff
@@ -97,4 +97,19 @@ def _new_learning_node(self, initial_stats=None, parent=None):
...
+                              binary_split=self.binary_split,
+                              stop_mem_management=self.stop_mem_management,
+                              remove_poor_attrs=self.remove_poor_attrs,
+                              merit_preprune=self.merit_preprune,
+                              leaf_prediction=self.leaf_prediction,
+                              nb_threshold=self.nb_threshold,
+                              nominal_attributes=self.nominal_attributes,
+                              max_features=self.max_features,
+                              max_depth=self.max_depth,
+                              seed=self._rng)
```
Comment (smastelini) [inline river/tree/arf_hoeffding_tree_classifier.py:115]: Was not it possible to devise a more automated solution with variants of `_get_params`?

File: river/ensemble/adaptive_random_forest.py [hunk 5]
```diff
@@ -0,0 +1,824 @@
...
+        self.drift_detector = drift_detector
+        self.warning_detector = warning_detector
+        self.seed = seed
+        self._rng = check_random_state(self.seed)   # Actual random number generator
+
+        # Internal parameters
+        self._n_samples_seen = 0
+        self._base_member_class = None
+
+    def learn_one(self, x: dict, y: base.typing.ClfTarget, **kwargs):
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:51]: Should we include typing annotation for the classification case here too?

Comment (smastelini) [reply river/ensemble/adaptive_random_forest.py:51]: Something like the generic `base.typing.Target`?

File: river/ensemble/adaptive_random_forest.py [hunk 6]
```diff
@@ -0,0 +1,824 @@
...
+                base_drift_detector=self.drift_detector,
+                base_warning_detector=self.warning_detector,
+                is_background_learner=False,
+                base_metric=copy.deepcopy(self.metric))
+            for i in range(self.n_models)
+        ]
+
+    @abc.abstractmethod
+    def _get_base_model(self, seed: int):
+        raise NotImplementedError
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:91]: Wonderful!

File: river/ensemble/adaptive_random_forest.py [hunk 7]
```diff
@@ -0,0 +1,824 @@
...
+        self.nb_threshold = nb_threshold
+        self.nominal_attributes = nominal_attributes
+        self.max_depth = max_depth
+
+    def _multiclass(self):
+        return True
+
+    def predict_proba_one(self, x: dict) -> typing.Dict[base.typing.ClfTarget, float]:
+
+        y_pred = collections.Counter()
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:300]: Nice workaround!

File: river/ensemble/adaptive_random_forest.py [hunk 8]
```diff
@@ -0,0 +1,824 @@
...
+    Notice that this implementation is slightly different from the original
+    algorithm proposed in [^2]. The `HoeffdingTreeRegressor` is used as base
+    learner, instead of `FIMT-DD`. It also adds a new strategy to monitor the
+    incoming data and check for concept drifts. The monitored data (either the
+    trees' errors or their predictions) are centered and scaled (z-score
+    normalization) to have zero mean and unit standard deviation. Transformed
+    values are then again normalized in the [0, 1] range to fulfil ADWIN's
+    requirements. We assume that the data subjected to the z-score
+    normalization lies within the interval of the mean $\pm3\sigma$, as it
+    occurs in normal distributions.
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:358]: Something along these lines (to reflect the new changes):

> The monitored data (either the trees' errors or their predictions) are normalized in the [0, 1] range to fulfil ADWIN's requirements. We assume that the data subjected to the normalization, e.g., the monitored error, follows a normal distribution, and thus, lies within the interval of the mean $\pm3\sigma$.

File: river/ensemble/adaptive_random_forest.py [hunk 9]
```diff
@@ -0,0 +1,824 @@
...
+                    weights = weights / weights.sum()
+                    y_pred *= weights
+        else:
+            for idx, model in enumerate(self.models):
+                y_pred[idx] = model.predict_one(x)
+
+        if self.aggregation_method == self._MEAN:
+            y_pred = y_pred.mean()
+        else:
+            y_pred = np.median(y_pred)
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:566]: Nice solution!

File: river/ensemble/adaptive_random_forest.py [hunk 10]
```diff
@@ -0,0 +1,824 @@
...
+    The main purpose of this class is to train the foreground model.
+    Optionally, it monitors drift detection. Depending on the configuration,
+    if drift is detected then the foreground model is reset or replaced by a
+    background model.
+
+    Parameters
+    ----------
+    index_original
+        Tree index within the ensemble.
+    base_model: ARFHoeffdingTreeClassifier
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:607]: Please, remove the type

File: river/ensemble/adaptive_random_forest.py [hunk 11]
```diff
@@ -0,0 +1,824 @@
...
+            self.model = self.background_learner.model
+            self.warning_detector = self.background_learner.warning_detector
+            self.drift_detector = self.background_learner.drift_detector
+            self.metric = copy.deepcopy(self.background_learner.base_metric)
+            self.created_on = self.background_learner.created_on
+            self.background_learner = None
+        else:
+            # Reset model
+            self.model = copy.deepcopy(self.base_model)
+            self.metric = copy.deepcopy(self.base_metric)
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:677]: We need to ensure that deepcopy works as we expect: clone the object without the modified internal state.

File: river/ensemble/adaptive_random_forest.py [hunk 12]
```diff
@@ -0,0 +1,824 @@
...
+            if self._use_background_learner:
+                self.warning_detector.update(drift_detector_input)
+                # Check if there was a (warning) change
+                if self.warning_detector.change_detected:
+                    self.last_warning_on = n_samples_seen
+                    self.n_warnings_detected += 1
+                    # Create a new background learner object
+                    self.background_learner = self.__class__(
+                        index_original=self.index_original,
+                        base_model=self.model.new_instance(),
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:705]: Cool!

File: river/ensemble/adaptive_random_forest.py [hunk 13]
```diff
@@ -0,0 +1,824 @@
...
+            self.created_on = self.background_learner.created_on
+            self.background_learner = None
+        else:
+            # Reset model
+            self.model = copy.deepcopy(self.base_model)
+            self.metric = copy.deepcopy(self.base_metric)
+            self.created_on = n_samples_seen
+            self.drift_detector = copy.deepcopy(self.base_drift_detector)
+
+    def learn_one(self, x: dict, y: base.typing.ClfTarget, *, sample_weight: int,   # noqa
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:681]: Change to `base.typing.Target`

File: river/ensemble/adaptive_random_forest.py [hunk 14]
```diff
@@ -0,0 +1,824 @@
...
+            if self.drift_detector.change_detected:
+                self.last_drift_on = n_samples_seen
+                self.n_drifts_detected += 1
+                self.reset(n_samples_seen)
+
+    @abc.abstractmethod
+    def _drift_detector_input(self,
+                              y_true: typing.Union[base.typing.ClfTarget, base.typing.RegTarget],
+                              y_pred: typing.Union[base.typing.ClfTarget, base.typing.RegTarget]):
+        raise NotImplementedError
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:728]: This is looking good!

File: river/ensemble/adaptive_random_forest.py [hunk 15]
```diff
@@ -0,0 +1,824 @@
...
+    """Forest member class for regression"""
+
+    def __init__(self,
+                 index_original: int,
+                 base_model: ARFHoeffdingTreeRegressor,
+                 created_on: int,
+                 base_drift_detector: base.DriftDetector,
+                 base_warning_detector: base.DriftDetector,
+                 is_background_learner,
+                 base_metric: typing.Union[MAE, MSE]):
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:772]: According to the docstring in ARFReg, the metric is an instance of any Regression metric. Is it right?

File: river/ensemble/adaptive_random_forest.py [hunk 16]
```diff
@@ -0,0 +1,824 @@
...
+            index_original=index_original,
+            base_model=base_model,
+            created_on=created_on,
+            base_drift_detector=base_drift_detector,
+            base_warning_detector=base_warning_detector,
+            is_background_learner=is_background_learner,
+            base_metric=base_metric
+        )
+        # Normalization of info monitored by drift detectors (using Welford's algorithm)
+        self._k = 0
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:783]: \# TODO: update this portion, as previously discussed!

Comment (smastelini) [reply river/ensemble/adaptive_random_forest.py:783]: Please, take `river.tree._nodes.hatr_nodes.py` as a reference.

**Comment by smastelini** [comment]
Wonderful work, @jacobmontiel! I finished a first pass in the code, and it is looking pretty good!

I really enjoyed the restructuring you did in the internal organization of ARF. It is looking much simpler and clear now!

We are almost there :D

Please, let me know if you need any assistance.

**Review by smastelini — CHANGES_REQUESTED** [review]
Wonderful work, @jacobmontiel! I finished a first pass in the code, and it is looking pretty good!

I really enjoyed the restructuring you did in the internal organization of ARF. It is looking much simpler and clear now!

We are almost there :D

Please, let me know if you need any assistance.

File: river/ensemble/__init__.py [hunk 17]
```diff
@@ -7,6 +8,7 @@
 
 
 __all__ = [
+    'AdaptiveRandomForestClassifier',
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/ensemble/__init__.py:13]: I think we are in the borderline w.r.t. names of other ensemble methods

```python
__all__ = [
    'AdaptiveRandomForestClassifier',
    'AdaptiveRandomForestRegressor',
    'AdaBoostClassifier',
    'ADWINBaggingClassifier',
    'BaggingClassifier',
    'BaggingRegressor',
    'LeveragingBaggingClassifier',
    'SRPClassifier'
]
```

File: river/test_.py (hunk 3 above)
Comment (jacobmontiel) [reply river/test_.py:40]: Personally, I prefer this convention as it is clear what it means.

This goes down to the mental load on the user, are these acronyms well know in the community? A counterexample is the `PAClassifier`. This is especially tricky with new methods. 

I don't oppose using `ARFClassifier` or `ARFRegressor` but is it really a big deal?

To a lesser degree, we can also account for the pronunciation of an acronym, `ADWIN`has a nice ring to it, on the other hand, `ARF` just sounds weird to me.

File: river/tree/arf_hoeffding_tree_classifier.py (hunk 4 above)
Comment (jacobmontiel) [reply river/tree/arf_hoeffding_tree_classifier.py:115]: I tried but it seems that `**kwargs` are trickier to retrieve. I suggest we use this solution for the time being while we devise a solution that can be applied to any estimator (also requiered for a `clone` method).

File: river/ensemble/adaptive_random_forest.py (hunk 9 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py:566]: From this comment, I spotted a bug in the code. It is fixed now. Thanks... I guess :sweat_smile:

File: river/ensemble/adaptive_random_forest.py (hunk 14 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py:728]: In future releases, this could be used to modify the classifier to track drift based on different metrics :-)

File: river/ensemble/adaptive_random_forest.py (hunk 15 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py:772]: I added a note as we have only tested with MAE and MSE.

File: river/ensemble/test_adaptive_random_forest.py (hunk 1 above)
Comment (jacobmontiel) [reply river/ensemble/test_adaptive_random_forest.py]: Added in 53c75bb9b45bfa0539d0c4bcd86bd3707603c5d9

File: river/ensemble/adaptive_random_forest.py (hunk 5 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py]: Nice catch, I changed it to `base.typing.Target` in 53c75bb9b45bfa0539d0c4bcd86bd3707603c5d9

File: river/ensemble/adaptive_random_forest.py (hunk 8 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py]: Updated the documentation in 53c75bb9b45bfa0539d0c4bcd86bd3707603c5d9

File: river/ensemble/adaptive_random_forest.py (hunk 9 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py]: Fix is in 53c75bb9b45bfa0539d0c4bcd86bd3707603c5d9

File: river/ensemble/adaptive_random_forest.py (hunk 10 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py]: Done in 53c75bb9b45bfa0539d0c4bcd86bd3707603c5d9

File: river/ensemble/adaptive_random_forest.py (hunk 11 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py:682]: `deepcopy` will create an exact copy of the object including its internal state. This is why I have to keep `base_*` objects to "clone". The actual updated objects are only names `*`, e.g. `base_metric` -> `metric`

File: river/ensemble/adaptive_random_forest.py (hunk 16 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py]: Updated in 53c75bb9b45bfa0539d0c4bcd86bd3707603c5d9

File: river/ensemble/adaptive_random_forest.py (hunk 8 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py]: Updated again in b628832e1062fda4cdbb7ae1c1948d4d6212bd90 to account for the updates to the drift detection strategy.

File: river/ensemble/adaptive_random_forest.py (hunk 13 above)
Comment (jacobmontiel) [reply river/ensemble/adaptive_random_forest.py]: Changed in b628832e1062fda4cdbb7ae1c1948d4d6212bd90

File: river/ensemble/__init__.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/ensemble/__init__.py:13]: Let's keep the current names.

File: river/test_.py (hunk 3 above)
Comment (smastelini) [reply river/test_.py:40]: Yes, it does sound weird hahaha

Well, the best solution is keeping the current names.

File: river/tree/arf_hoeffding_tree_classifier.py (hunk 4 above)
Comment (smastelini) [reply river/tree/arf_hoeffding_tree_classifier.py]: I agree. This is just a minor detail :)

File: river/ensemble/adaptive_random_forest.py (hunk 11 above)
Comment (smastelini) [reply river/ensemble/adaptive_random_forest.py:849]: Understood! That's a nice workaround. Once we have proper cloning capabilities we can simplify that.

File: river/ensemble/adaptive_random_forest.py (hunk 15 above)
Comment (smastelini) [reply river/ensemble/adaptive_random_forest.py]: Perfect!

File: river/ensemble/adaptive_random_forest.py [hunk 18]
```diff
@@ -0,0 +1,977 @@
...
+                              merit_preprune=self.merit_preprune,
+                              leaf_prediction=self.leaf_prediction,
+                              nb_threshold=self.nb_threshold,
+                              nominal_attributes=self.nominal_attributes,
+                              max_features=self.max_features,
+                              max_depth=self.max_depth,   # noqa
+                              seed=self._rng)
+
+
+class BaseTreeRegressor(HoeffdingTreeRegressor):
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:210]: I like these new names!

File: river/ensemble/adaptive_random_forest.py [hunk 19]
```diff
@@ -0,0 +1,977 @@
...
+    3. drift detectors per base tree, which cause selective resets in response
+       to drifts.
+
+    Notice that this implementation is slightly different from the original
+    algorithm proposed in [^2]. The `HoeffdingTreeRegressor` is used as base
+    learner, instead of `FIMT-DD`. It also adds a new strategy to monitor the
+    predictions and check for concept drifts. The monitored is are normalized
+    in the [0, 1] range to fulfil ADWIN's requirements. We assume that the data
+    subjected to the normalization follows a normal distribution, and thus,
+    lies within the interval of the mean $\pm3\sigma$.
```
Comment (smastelini) [inline river/ensemble/adaptive_random_forest.py:524]:
```suggestion
    Notice that this implementation is slightly different from the original
    algorithm proposed in [^2]. The `HoeffdingTreeRegressor` is used as base
    learner, instead of `FIMT-DD`. It also adds a new strategy to monitor the
    predictions and check for concept drifts. The deviations of the predictions
    to the target are monitored and normalized in the [0, 1] range to fulfill ADWIN's
    requirements. We assume that the data subjected to the normalization follows
    a normal distribution, and thus, lies within the interval of the mean $\pm3\sigma$.
```

**Review by smastelini — APPROVED** [review]

**Comment by smastelini** [comment]
Thanks for all the nice work, @jacobmontiel! This PR brought a lot of improvements to the current structure.
